README.md
JAPAN - Power Usage Checker Chrome Extension
============================================

< [Chrome Webstore](https://chrome.google.com/webstore/detail/hclgegipaehbigmbhdpfapmjadbaldib) >

東京電力（TEPCO）, 東北電力, 関西電力, 九州電力, 中部電力, 北海道電力の電力使用量を表示することができます。  
東北地方太平洋沖地震の影響により、電気の供給が不足しています。  
この拡張機能で現在の電力の使用状況を確認し、節電に努めていただければ幸いです。

ツールバーには選択した電力会社の電力使用率を表示します。
アイコンをクリックすると更に詳細な情報をみることができます。

※この拡張機能は API 提供元が提供を停止した際に削除いたします。
※正しい情報ではない可能性も御座います。予めご了承ください。


対応電力会社
------------
- 東京電力
- 東北電力
- 関西電力
- 九州電力
- 中部電力
- 北海道電力


使用 API
--------

電力使用状況 API: http://developer.yahoo.co.jp/webapi/shinsai/
※ 2013/05/11 からこちらのAPIを利用しています。

Go 節電 API: http://api.gosetsuden.jp/
※ 2011/08/25 ~ 2013/05/10 までこちらの API を利用していました。

東京電力電力供給状況 API: http://tepco-usage-api.appspot.com/
※ 2011/08/24 までこちらの API を利用していました。


AUTHOR
------

Tetsuwo OISHI - 
tetsuwo.oishi@gmail.com / 
[Facebook](http://fb.me/tetsuwo) /
[Twitter](http://twitter.com/tetsukamp)





CHANGELOG.md
CHANGE LOG:
-----------

- 2.3.0-2.3.3
  - Go 節電 API の提供停止のため、API の乗り換え
  - API へのアクセス数を軽減するため
- 2.2.2
  - CSP により、インラインスクリプトでエラーが出ていた不具合を修正 
- 2.2.1
  - 北海道電力、北陸電力、中国電力、四国電力 に対応 
- 2.2.0
  - Manifest Version 2 に対応 
- 2.1.0
  - 推移ページ作成 
- 2.0.3
  - 細部修正
- 2.0.2
  - 細部修正
- 2.0.0
  - 東北電力、中部電力、関西電力、九州電力に対応しました
- 1.0.4
  - バグ対応
- 1.0.3
  - バグ対応
- 1.0.2
  - 更新日時を表示
- 0.0.1
  - 開発開始



src/js/page-popup.js

function DOMId(id) {
    return document.getElementById(id);
}

function togglePage(page) {
    switch (page) {
        case 'info':
            region = JSON.parse(localStorage.region);
            update();
            DOMId('epco-name').innerHTML = chrome.i18n.getMessage(region);
            DOMId('info').style.display = 'block';
            DOMId('setting').style.display = 'none';
            DOMId('btn-info').disabled = true;
            DOMId('btn-setting').disabled = false;
            break;
        case 'setting':
            region = JSON.parse(localStorage.region);
            for (var i = 0; i < DOMId('epco').options.length; i++) {
                if (DOMId('epco').options[i].value === region) {
                    DOMId('epco').options[i].selected = true;
                    break;
                }
            }
            DOMId('info').style.display = 'none';
            DOMId('setting').style.display = 'block';
            DOMId('btn-info').disabled = false;
            DOMId('btn-setting').disabled = true;
            break;
    }
}

function update() {
    var usageinfo = JSON.parse(localStorage.usageinfo);

    DOMId('graph-hide').style.width = (100 - parseInt(usageinfo.result)) + '%';
    DOMId('data-usage').innerHTML = usageinfo.instant.usage.toString().replace(/([0-9]+?)(?=(?:[0-9]{3})+$)/g, '$1,');
    DOMId('data-capacity').innerHTML = usageinfo.peak.usage.toString().replace(/([0-9]+?)(?=(?:[0-9]{3})+$)/g, '$1,');
    DOMId('data-userate').innerHTML = usageinfo.result;

    var explodeDate = usageinfo.updated.date.split('-');
    console.log(explodeDate);
    var t = {
        Y: explodeDate[0],
        m: explodeDate[1],
        d: explodeDate[2],
        H: usageinfo.updated.hour
    };
    var tmp = chrome.i18n.getMessage('usage_updated_format').replace('Y', t.Y).replace('m', t.m).replace('d', t.d).replace('H', t.H);
    DOMId('data-usage_updated').innerHTML = tmp;

    DOMId('link1').setAttribute('href', chrome.i18n.getMessage(region + '_url'));
    DOMId('link2').setAttribute('href', 'http://setsuden.yahoo.co.jp/' + region + '/');
    DOMId('link').style.display = 'block';
}

DOMId('save').onclick = function() {
    if (!DOMId('epco').value) {
        DOMId('error').innerHTML = 'Please select!';
        return false;
    }
    DOMId('error').innerHTML = '';
    localStorage.region = JSON.stringify(DOMId('epco').value);
    JapanPowerUsage.run(true);
};

DOMId('btn-info').onclick = function() {
    togglePage('info');
};

DOMId('btn-setting').onclick = function() {
    togglePage('setting');
};

//DOMId('btn-transition').onclick = function() {
//    window.open('transition.html');
//};

i18n();

if (localStorage.region) {
    var region = JSON.parse(localStorage.region);
    togglePage('info');
} else {
    var region = 'tokyo';
    localStorage.region = JSON.stringify(region);
    togglePage('setting');
    JapanPowerUsage.run(true);
}

if (localStorage.usageinfo) {
    update();
}



src/js/raphael-plugins.js
(function() {
var tokenRegex = /\{([^\}]+)\}/g,
    objNotationRegex = /(?:(?:^|\.)(.+?)(?=\[|\.|$|\()|\[('|")(.+?)\2\])(\(\))?/g, // matches .xxxxx or ["xxxxx"] to run over object properties
    replacer = function (all, key, obj) {
        var res = obj;
        key.replace(objNotationRegex, function (all, name, quote, quotedName, isFunc) {
            name = name || quotedName;
            if (res) {
                if (name in res) {
                    res = res[name];
                }
                typeof res == "function" && isFunc && (res = res());
            }
        });
        res = (res === null || res == obj ? all : res) + "";
        return res;
    },
    fill = function (str, obj) {
        return String(str).replace(tokenRegex, function (all, key) {
            return replacer(all, key, obj);
        });
    };
    Raphael.fn.popup = function (X, Y, set, pos, ret) {
        pos = String(pos || "top-middle").split("-");
        pos[1] = pos[1] || "middle";
        var r = 5,
            bb = set.getBBox(),
            w = Math.round(bb.width),
            h = Math.round(bb.height),
            x = Math.round(bb.x) - r,
            y = Math.round(bb.y) - r,
            gap = Math.min(h / 2, w / 2, 10),
            shapes = {
                top: "M{x},{y}h{w4},{w4},{w4},{w4}a{r},{r},0,0,1,{r},{r}v{h4},{h4},{h4},{h4}a{r},{r},0,0,1,-{r},{r}l-{right},0-{gap},{gap}-{gap}-{gap}-{left},0a{r},{r},0,0,1-{r}-{r}v-{h4}-{h4}-{h4}-{h4}a{r},{r},0,0,1,{r}-{r}z",
                bottom: "M{x},{y}l{left},0,{gap}-{gap},{gap},{gap},{right},0a{r},{r},0,0,1,{r},{r}v{h4},{h4},{h4},{h4}a{r},{r},0,0,1,-{r},{r}h-{w4}-{w4}-{w4}-{w4}a{r},{r},0,0,1-{r}-{r}v-{h4}-{h4}-{h4}-{h4}a{r},{r},0,0,1,{r}-{r}z",
                right: "M{x},{y}h{w4},{w4},{w4},{w4}a{r},{r},0,0,1,{r},{r}v{h4},{h4},{h4},{h4}a{r},{r},0,0,1,-{r},{r}h-{w4}-{w4}-{w4}-{w4}a{r},{r},0,0,1-{r}-{r}l0-{bottom}-{gap}-{gap},{gap}-{gap},0-{top}a{r},{r},0,0,1,{r}-{r}z",
                left: "M{x},{y}h{w4},{w4},{w4},{w4}a{r},{r},0,0,1,{r},{r}l0,{top},{gap},{gap}-{gap},{gap},0,{bottom}a{r},{r},0,0,1,-{r},{r}h-{w4}-{w4}-{w4}-{w4}a{r},{r},0,0,1-{r}-{r}v-{h4}-{h4}-{h4}-{h4}a{r},{r},0,0,1,{r}-{r}z"
            },
            offset = {
                hx0: X - (x + r + w - gap * 2),
                hx1: X - (x + r + w / 2 - gap),
                hx2: X - (x + r + gap),
                vhy: Y - (y + r + h + r + gap),
                "^hy": Y - (y - gap)

            },
            mask = [{
                x: x + r,
                y: y,
                w: w,
                w4: w / 4,
                h4: h / 4,
                right: 0,
                left: w - gap * 2,
                bottom: 0,
                top: h - gap * 2,
                r: r,
                h: h,
                gap: gap
            }, {
                x: x + r,
                y: y,
                w: w,
                w4: w / 4,
                h4: h / 4,
                left: w / 2 - gap,
                right: w / 2 - gap,
                top: h / 2 - gap,
                bottom: h / 2 - gap,
                r: r,
                h: h,
                gap: gap
            }, {
                x: x + r,
                y: y,
                w: w,
                w4: w / 4,
                h4: h / 4,
                left: 0,
                right: w - gap * 2,
                top: 0,
                bottom: h - gap * 2,
                r: r,
                h: h,
                gap: gap
            }][pos[1] == "middle" ? 1 : (pos[1] == "top" || pos[1] == "left") * 2];
            var dx = 0,
                dy = 0,
                out = this.path(fill(shapes[pos[0]], mask)).insertBefore(set);
            switch (pos[0]) {
                case "top":
                    dx = X - (x + r + mask.left + gap);
                    dy = Y - (y + r + h + r + gap);
                    break;
                case "bottom":
                    dx = X - (x + r + mask.left + gap);
                    dy = Y - (y - gap);
                    break;
                case "left":
                    dx = X - (x + r + w + r + gap);
                    dy = Y - (y + r + mask.top + gap);
                    break;
                case "right":
                    dx = X - (x - gap);
                    dy = Y - (y + r + mask.top + gap);
                    break;
                default: break;
            }
            out.translate(dx, dy);
            if (ret) {
                ret = out.attr("path");
                out.remove();
                return {
                    path: ret,
                    dx: dx,
                    dy: dy
                };
            }
            set.translate(dx, dy);
            return out;
    };
})();

Raphael.fn.drawGrid = function (x, y, w, h, wv, hv, color) {
    color = color || "#000";
    var path = ["M", Math.round(x) + 0.5, Math.round(y) + 0.5,
		"L", Math.round(x + w) + 0.5, Math.round(y) + 0.5,
		Math.round(x + w) + 0.5, Math.round(y + h) + 0.5,
		Math.round(x) + 0.5, Math.round(y + h) + 0.5,
		Math.round(x) + 0.5, Math.round(y) + 0.5],
        rowHeight = h / hv,
        columnWidth = w / wv;
    for (var i = 1; i < hv; i++) {
        path = path.concat(["M", Math.round(x) + 0.5, Math.round(y + i * rowHeight) + 0.5,
			    "H", Math.round(x + w) + 0.5]);
    }
    for (i = 1; i < wv; i++) {
        path = path.concat(["M", Math.round(x + i * columnWidth) + 0.5, Math.round(y) + 0.5,
			    "V", Math.round(y + h) + 0.5]);
    }
    return this.path(path.join(",")).attr({stroke: color});
};


src/js/japan-power-usage.js
JapanPowerUsage = new function() {
    var toolbar = chrome.browserAction;
    var api = 'http://setsuden.yahooapis.jp';
    var data = {
        peak: {},
        instant: {},
        updated: { date: null, hour: null }
    };

    function indicate() {
        data.ratio = (data.instant.usage * 100) / data.peak.usage;
        data.result = Math.round(data.ratio);
        toolbar.setIcon({ path: 'images/meter_'+ String(Math.ceil(data.result/10)) +'.png' });
        toolbar.setBadgeText({ text: String(data.result) +'%' });
        localStorage.usageinfo = JSON.stringify(data);
    }

    function serialize(param, prefix) {
        var query = [];
        for(var p in param) {
            var k = prefix ? prefix + '[' + p + ']' : p, v = param[p];
            query.push(
                typeof v == 'object' ?
                    this.serialize(v, k) :
                    encodeURIComponent(k) + '=' + encodeURIComponent(v)
            );
        }
        return query.join('&');
    }

    function get() {
        var region = JSON.parse(localStorage.region);

        var param = {};
        param.appid = 'dj0zaiZpPW0xdDdMOFp1MDRudCZkPVlXazljMEYwV2xrNU5EZ21jR285TUEtLSZzPWNvbnN1bWVyc2VjcmV0Jng9Yjk-';
        param.area = region;
        param.output = 'json';

        var req = new XMLHttpRequest();
        req.open('GET', api + '/v1/Setsuden/latestPowerUsage?' + serialize(param), true);
        req.onreadystatechange = function() {
            if (req.readyState === 4) {
                var res = JSON.parse(req.responseText);
                if (res.ElectricPowerUsage) {
                    data.peak.usage = res.ElectricPowerUsage.Capacity.$;
                    data.instant.usage = res.ElectricPowerUsage.Usage.$;
                    data.updated.date = res.ElectricPowerUsage.Date;
                    data.updated.hour = res.ElectricPowerUsage.Hour;
                    indicate();
                }
            }
        }
        req.send(null);
    }

    this.log = function(a) {
        console.log(a);
    };

    this.run = function(init) {
        if (init) {
            toolbar.setBadgeBackgroundColor({ color: [66, 66, 66, 255] });
            toolbar.setBadgeText({ text: '-- %' });
        }
        get();
    };

    this.getData = function() {
        return data;
    };
};


src/js/tepco-power-usage.js
TEPCOPowerUsage = new function()
{
    var API_URL = 'http://tepco-usage-api.appspot.com/latest.json',
        data = {},
        toolbar = chrome.browserAction;

    function indicate()
    {
        data.ratio = (data.usage * 100) / data.capacity;
        data.result = Math.round(data.ratio);
        toolbar.setIcon({path:'img/meter_' + String(Math.ceil(data.result/10)) + '.png'});
        toolbar.setBadgeText({text:String(data.result)+'%'});
		localStorage.usageinfo = JSON.stringify(data);
    }

    function get()
    {
        var req = new XMLHttpRequest();
        req.open('GET', API_URL, true);
        req.onreadystatechange = function()
        {
            if (req.readyState === 4)
            {
                data = JSON.parse(req.responseText);
                indicate();
            }
        }
        req.send(null);
    }

    this.log = function(a)
    {
        console.log(a);
    };

    this.run = function(init)
    {
		if (init)
		{
	        toolbar.setBadgeBackgroundColor({color:[66, 66, 66, 255]});
	        toolbar.setBadgeText({text:'-- %'});
	    }
        get();
    };

    this.getData = function()
    {
		return data;
    };
};

src/js/page-transition.js

function makeChart(labels, data) {
    $('#chart').empty();

    var design = {
            label: {
                font: 'bold 11px Helvetica, Arial',
                fill: '#777'
            }
            , date: {
                font: 'bold 12px Helvetica, Arial',
                fill: '#fff'
            }
            , data: {
                font: 'bold 13px Helvetica, Arial',
                fill: '#fff'
            }
        },
        gutter = { top: 20, bottom: 20, left: 0 },
        MAX = Math.max.apply(Math, data),
        W = 600,
        H = 300,
        R = Raphael('chart', W, H),
        C = '#8c0',
        X = (W - gutter.left) / labels.length,
        Y = (H - gutter.bottom - gutter.top) / MAX;

    $('#actual-max').text(setComma(MAX));

    R.drawGrid(
        gutter.left + X * 0.5 + 0.5,
        gutter.top + 0.5,
        W - gutter.left - X,
        H - gutter.top - gutter.bottom,
        10, 10, '#333'
    );

    var path = R.path().attr({stroke: C, 'stroke-width': 4, 'stroke-linejoin': 'round'}),
        bgp = R.path().attr({stroke: 'none', opacity: 0.3, fill: C}),
        label = R.set(),
        is_label_visible = false,
        leave_timer,
        blanket = R.set();

    label.push(R.text(60, 12, '100,000 kW').attr(design.data));
    label.push(R.text(60, 27, '2011/3/11 15h').attr(design.date).attr({fill: C}));
    label.hide();

    var frame = R.popup(100, 100, label, 'right')
                 .attr({
		     fill: '#000',
		     stroke: '#666',
		     'stroke-width': 2,
		     'fill-opacity': 0.7
		 }).hide();

    var p, bgpp;
    for (var i = 0, len = labels.length; i < len; i++) {
        var time = new Date(labels[i]),
            y = Math.round(H - gutter.bottom - Y * (data[i])),
            x = Math.round(gutter.left + X * (i + 0.5)),
            t = R.text(x, H - 6, time.getHours()).attr(design.label).toBack();
        if (!i) {
            p = ['M', x, y, 'C', x, y];
            bgpp = ['M', gutter.left + X * 0.5, H - gutter.bottom, 'L', x, y, 'C', x, y];
        }
        if (i && i < len - 1) {
            var Y0 = Math.round(H - gutter.bottom - Y * data[i - 1]),
                X0 = Math.round(gutter.left + X * (i - 0.5)),
                Y2 = Math.round(H - gutter.bottom - Y * data[i + 1]),
                X2 = Math.round(gutter.left + X * (i + 1.5));
            var a = getAnchors(X0, Y0, x, y, X2, Y2);
            p = p.concat([a.x1, a.y1, x, y, a.x2, a.y2]);
            bgpp = bgpp.concat([a.x1, a.y1, x, y, a.x2, a.y2]);
        }
        var dot = R.circle(x, y, 4).attr({fill: '#111', stroke: C, 'stroke-width': 2});
        blanket.push(R.rect(gutter.left + X * i, 0, X, H - gutter.bottom).attr({stroke: 'none', fill: '#fff', opacity: 0}));
        var rect = blanket[blanket.length - 1];

        (function(x, y, d, t, dot) {
            var timer, i = 0;
            rect.hover(function() {
                clearTimeout(leave_timer);
                var side = (x + frame.getBBox().width > W) ? 'left' : 'right';
                var ppp = R.popup(x, y, label, side, 1);
                frame.show().stop().animate({path: ppp.path}, 200 * is_label_visible);
                label[0].attr({text: setComma(d) + ' kW'}).show().stop().animateWith(frame, {translation: [ppp.dx, ppp.dy]}, 200 * is_label_visible);
                label[1].attr({text: getYmdH(t)}).show().stop().animateWith(frame, {translation: [ppp.dx, ppp.dy]}, 200 * is_label_visible);
                dot.attr('r', 6);
                is_label_visible = true;
            }, function() {
                dot.attr('r', 4);
                leave_timer = setTimeout(function () {
                    frame.hide();
                    label[0].hide();
                    label[1].hide();
                    is_label_visible = false;
                }, 1);
            });
        })(x, y, data[i], time, dot);
    }
    p = p.concat([x, y, x, y]);
    bgpp = bgpp.concat([x, y, x, y, 'L', x, H - gutter.bottom, 'z']);
    path.attr({path: p});
    bgp.attr({path: bgpp});
    frame.toFront();
    label[0].toFront();
    label[1].toFront();
    blanket.toFront();
}

var regions = {
    tokyo:  { name: '東京電力', min: '2007-01-01' },
    tohoku: { name: '東北電力', min: '2008-04-01' },
    kansai: { name: '関西電力', min: '2011-08-03' },
    kyushu: { name: '九州電力', min: false },
    chubu:  { name: '中部電力', min: '2011-08-08' }
};

function getAnchors(p1x, p1y, p2x, p2y, p3x, p3y) {
    var l1 = (p2x - p1x) / 2,
        l2 = (p3x - p2x) / 2,
        a = Math.atan((p2x - p1x) / Math.abs(p2y - p1y)),
        b = Math.atan((p3x - p2x) / Math.abs(p2y - p3y));
    a = p1y < p2y ? Math.PI - a : a;
    b = p3y < p2y ? Math.PI - b : b;
    var alpha = Math.PI / 2 - ((a + b) % (Math.PI * 2)) / 2,
        dx1 = l1 * Math.sin(alpha + a),
        dy1 = l1 * Math.cos(alpha + a),
        dx2 = l2 * Math.sin(alpha + b),
        dy2 = l2 * Math.cos(alpha + b);
    return {
        x1: p2x - dx1,
        y1: p2y + dy1,
        x2: p2x + dx2,
        y2: p2y + dy2
    };
}

function getYmd() {
    var t = new Date(), month = t.getMonth() + 1;
    return t.getFullYear() + '-'
        + (month < 10 ? '0' + month : month) + '-'
        + (t.getDate() < 10 ? '0'+ t.getDate() : t.getDate());
}

function getYmdH(t) {
    return t.getFullYear() + '/' + (t.getMonth() + 1) + '/' + t.getDate() + ' ' + t.getHours() + 'h';
}

function setComma(n) {
    var r = new String(n).replace(/,/g, '');
    while (r != (r = r.replace(/^(-?\d+)(\d{3})/, '$1,$2')));
    return r;
}

function changeRegion() {
    $('#date').attr('min', regions[$('#region').val()].min).attr('max', getYmd());
}

function loadUsage() {
    $('#nowloading').show();
    $('#chart').hide();
    $('#date').val() || $('#date').val(getYmd());
    $.ajax({
        url: 'http://api.gosetsuden.jp/usage/' + $('#region').val() + '/actual',
        type: 'get',
        data: { date: $('#date').val() },
        dataType: 'json',
        success: function(json) {
            var labels = [], data = [];
            for (var row in json) {
                labels.push(json[row].timestamp);
                data.push(json[row].usage);
            }
            makeChart(labels, data);
            $('#nowloading').hide();
            $('#chart').show();
        }
    });
    $.ajax({
        url: 'http://api.gosetsuden.jp/peak/' + $('#region').val() + '/demand',
        type: 'get',
        data: { date: $('#date').val() },
        dataType: 'json',
        success: function(json) {
            $('#peak-demand').text(setComma(json[0].usage));
        }
    });
    $.ajax({
        url: 'http://api.gosetsuden.jp/peak/' + $('#region').val() + '/supply',
        type: 'get',
        data: { date: $('#date').val() },
        dataType: 'json',
        success: function(json) {
            $('#peak-supply').text(setComma(json[0].usage));
        }
    });
}

$('#region').change(changeRegion);
$('#load').click(loadUsage);
changeRegion();
loadUsage();

